permission-element.bs

```diff
@@ -19,7 +18,408 @@ Abstract: A `<permission>` HTML element to request browser permissions in-page.
+1. Append each [=powerful feature=] name to the {{[[Types]]}} [=ordered set=].
+
+</div>
+
+### Activation blockers ### {#permission-element-activation-blockers}
+
+<pre class=idl>
+enum PermissionElementInvalidReason {
+  "",  // No blocker reason.
+  "style", "type_count", "illegal_subframe", "covered",
+  "recently_moved", "recently_created"
+};
+</pre>
+
+Issue: Is covered and visible the same type of blocker?
+
+Issue: Are unsupported types also a blocker?
+
+<div class="issue">
+In the Chrome implementation, I find these reasons:
+    * type_invalid
+    * illegal_subframe
+    * unsuccesful_registration
+    * recently_attached
+    * intersection_visible
+    * intersection_changed
+    * intersection_out_of_viewport_or_clipped
+    * intersection_occluded_or_distorted
+    * style_invalid
+
+</div>
+
+The permission element keeps track of "blockers", reasons why the element (currently) cannot be activated. These blockers come with three lifetimes: Permanent, temporary, and expiring.
+
+: Permanent blocker
+:: Once an element has a permanent blocker, it will be disabled permanently.
+    There are used for issues that the website owner is expected to fix.
+    An example is a <{permission}> element inside a <{fencedframe}>.
+: Temporary blocker
+:: This is a blocker that will only be valid until the blocking condition no
+    no longer occurs. An example is a <{permission}> element that is not
+    currently in view.
+: Expiring blocker
+:: This is a blocker that is only valid for a fixed period of time. This is
+    used to block abuse scenarios like "click jacking". An example is
+    a <{permission}> element that has recently been moved.
+
+<div>
+<dfn dfn lt="blocker reason table"></dfn>
+<table class="def">
+<tr><th>style
+<td>permanent (???)
+<td>Something something.
+<tr><th>type_count
+<td>permanent
+<td>When another <{permission}> for the same [=powerful feature|permission type=] is used on the same page.
+<tr><th>illegal_subframe
+<td>permanent
+<td>When the <{permission}> is used inside a <{fencedframe}>.
+<tr><th>covered
+<td>temporary
+<td>When the <{permission}> is partially occluded, or not visible at all.
+<tr><th>recently_moved
+<td>expiring
+<td>When the <{permission}> is being moved on the screen.
+<tr><th>recently_created
+<td>expiring
+<td>When <{permission}> is attached to the document.
+</table>
+</div>
+
+<div algorithm>
+To <dfn for="HTMLPermissionElement">add a blocker</dfn> with a
+{{PermissionElementInvalidReason}} |reason| and an optional flag |expires|:
+
+1. [=Assert=]: |reason| is not `""`.
+    (The empty string in {{PermissionElementInvalidReason}} signals no blocker
+    is present. Why would you add a non-blocking blockern empty string?)
+1. Let |timestamp| be None.
+1. If |expires|, then let |timestamp| be [=current high resolution time=]
+    plus the [=blocker delay=].
+1. [=list/Append=] an entry to the internal {{[[BlockerList]]}} with |reason|
```

Comment:
Should we consider using a map and potentially replacing existing blockers that have the same 
reason (based on whichever timestamp is further in the future)? Or should we make a note for implementors to consider this, similarly to the other one?

Reply:
Not sure. Generally, I'd prefer to keep the spec simple, and let implementors do clever things if they wish, as long as they're spec compliant. Since the infra spec has good support for lists (in terms of pre-defined and reference-able spec algorithms), using lists at the spec level is both easy to write and easy to understand.